### PR TITLE
FPGA: Adding a New Sample: `memory_system`

### PR DESCRIPTION
**Pull request by haoyanwa**
# Adding a New Sample: `memory_system`

## Description

This code sample demonstrates how to implement an efficient banked memory system. We show the users how to partition the memory system, the effect that memory access has to the system performance, how to mitigate bad memory access/layout, and how the compiler optimizes memory banks.

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [x] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com
- [x] FPGA Regtest (Linux) https://spetc.intel.com/testanalysis?trview=0&testRunIds=10661780
- [x] FPGA Regtest (Windows) https://spetc.intel.com/testanalysis?trview=5&runBy=85740096&runBy_reporting_eng=true&testRunIds=10698400&tapgsize=1500

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [x] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [x] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/assets/waveform_naive.png
Comment (whitepau) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/assets/waveform_naive.png:1]: naive

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/assets/waveform_optimized.png
Comment (whitepau) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/assets/waveform_optimized.png:1]: optimized

**Review by whitepau — COMMENTED** [review]
alternative screenshots for waveforms that are a bit easier to read

**Review by whitepau — APPROVED** [review]

**Review by tomlenth — APPROVED** [review]
Another thorough readme.  thx.

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/README.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+   > **Note**: You can change the default target by using the command:
+   > ```
+   > cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
+   > ```
+
+3. Compile the design through the generated `Makefile`. The following build targets are provided, matching the recommended development flow:
+
+   | Compilation Type    | Command
+   |:---                 |:---
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/README.md:213]: This is not in line with the rest of the samples readmes

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/main.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#include <iostream>
+
+#include "exception_handler.hpp"
+#include "memory_system_defines.hpp"
+#include "memory_system_kernels.hpp"
+
+// Forward declaration of kernel names.
+class ID_NaiveKernel;
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/main.cpp:8]: There should be no `_` in class names
`IDNaiveKernel`

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/main.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+class ID_OptimizedKernel;
+
+// Function that verifies the output from example kernels.
+bool CheckOutput(sycl::queue &q) {
+  SimpleOutputT resultNaive;
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/main.cpp:13]: variable names should be snake_case: `result_naive`

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    } else {
+      std::cout << "Verification FAILED.\n";
+    }
+  }
+  catch (sycl::exception const &e) {
+    std::cerr << "Caught a synchronous SYCL exception: " << e.what()
+              << std::endl;
+    std::terminate();
+  }
+}
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/main.cpp:65]: no return?

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_defines.hpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+// Define input/output streaming interfaces
+/////////////////////////////////////////////
+using PipePropertiesT = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::bits_per_symbol<8>,
+    sycl::ext::intel::experimental::uses_valid<true>,
+    sycl::ext::intel::experimental::ready_latency<0>,
+    sycl::ext::intel::experimental::first_symbol_in_high_order_bits<true>));
+
+// Interfaces for the illustrative simple kernels.
+class ID_InStream_NaiveKernel;
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_defines.hpp:32]: CamelCase

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_defines.hpp [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+using InStream_NaiveKernel =
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_defines.hpp:33]: CamelCase

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_kernels.hpp [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        array2d[row][col] = array2d[row + 1][col];
+      }
+      array2d[kNumRows - 1][col] = input;
+      
+      #pragma unroll
+      for (int row = 0; row < kNumRows; ++row) {
+        output[row] = array2d[row][col];
+      }
+      
+      OutStream_NaiveKernel::write(output);
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_kernels.hpp:26]:
```suggestion
      OutputStreamNaiveKernel::write(output);
```

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_kernels.hpp (hunk 7 above)
Comment (yuguen-intel) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/DesignPatterns/banked_memory_system/src/memory_system_kernels.hpp:26]: Sorry I meant `OutSteamNaiveKernel`

**Review by yuguen-intel — APPROVED** [review]

**Pull request merged**